fv3core/grid/generation.py

```diff
@@ -0,0 +1,2136 @@
+        9---4---8
+        |       |
+        1   5   3
+        |       |
+        6---2---7
+        """
+        if self._sin_sg7 is None:
+            self._init_cell_trigonometry()
+        return self._sin_sg7
+
+    @property
+    def sin_sg8(self):
+        """
+        Sine of the angle at point 8 of the 'supergrid' that refines each grid cell:
+        9---4---8
+        |       |
+        1   5   3
+        |       |
+        6---2---7
+        """
+        if self._sin_sg8 is None:
+            self._init_cell_trigonometry()
+        return self._sin_sg8
+
+    @property
+    def sin_sg9(self):
+        """
+        Sine of the angle at point 9 of the 'supergrid' that refines each grid cell:
+        9---4---8
+        |       |
+        1   5   3
+        |       |
+        6---2---7
+        """
+        if self._sin_sg9 is None:
+            self._init_cell_trigonometry()
+        return self._sin_sg9
+
+    @property
+    def cosa(self):
+        """
+        The inner product of ee1 and ee2
+        """
+        if self._cosa is None:
+            self._init_cell_trigonometry()
+        return self._cosa
+
+    @property
+    def sina(self):
+        """
+        1-cosa**2
+        """
+        if self._sina is None:
+            self._init_cell_trigonometry()
+        return self._sina
+
+    @property
+    def cosa_u(self):
+        """
+        the average curve along the left cell-edge
+        as measured from the left and right sides
+        i.e. the average of cos_sg1[i,j] and cos_sg3[i-1, j]
+        """
+        if self._cosa_u is None:
+            self._init_cell_trigonometry()
+        return self._cosa_u
+
+    @property
+    def cosa_v(self):
+        """
+        the average curve along the bottom cell-edge
+        as measured from the left and right sides
+        i.e. the average of cos_sg2[i,j] and cos_sg4[i-1, j]
+        """
+        if self._cosa_v is None:
+            self._init_cell_trigonometry()
+        return self._cosa_v
+
+    @property
+    def cosa_s(self):
+        """
+        equivalent to cos_sg5, the inner product of ec1 and ec2
+        """
+        if self._cosa_s is None:
+            self._init_cell_trigonometry()
+        return self._cosa_s
+
+    @property
+    def sina_u(self):
+        """
+        the average curve along the left cell-edge
+        as measured from the left and right sides
+        i.e. the average of sin_sg1[i,j] and sin_sg3[i-1, j]
+        """
+        if self._sina_u is None:
+            self._init_cell_trigonometry()
+        return self._sina_u
+
+    @property
+    def sina_v(self):
+        """
+        the average curve along the bottom cell-edge
+        as measured from the left and right sides
+        i.e. the average of sin_sg2[i,j] and sin_sg4[i-1, j]
+        """
+        if self._sina_v is None:
+            self._init_cell_trigonometry()
+        return self._sina_v
+
+    @property
+    def rsin_u(self):
+        """
+        1/sina_u**2
```

Comment:
Note that the inverses are all defined as squares since they are only used as such.